components/canvas/canvas_paint_task.rs

```diff
@@ -85,4 +94,101 @@ impl CanvasPaintTask {
+        }
+        if dest_rect.size.width == 0 {
+            dest_rect.size.width = 1;
+        }
+        if dest_rect.size.height == 0 {
+            dest_rect.size.height = 1;
+        }
+
+        let canvas_rect = Rect(Point2D(0i32, 0i32), canvas_size);
+        let src_read_rect = canvas_rect.intersection(&dest_rect).unwrap_or(Rect::zero());
+
+        let mut dest_data = Vec::new();
+        //load the canvas data to the source vector
+        if !src_read_rect.is_empty() && canvas_size.width != 0 && canvas_size.height != 0 {
+            let data_surface = self.drawtarget.snapshot().get_data_surface();
+            let mut src_data = Vec::new();
+            data_surface.with_data(|element| {
+                src_data = element.to_vec();
+            });
+
+            let stride = data_surface.stride();
+
+            //start offset of the copyable rectangle
+            let mut src = (src_read_rect.origin.y * stride + src_read_rect.origin.x * 4) as usize;
+            //copy the data to the destination vector
+            for _ in range(0, src_read_rect.size.height) {
+                let row = src_data.slice(src, src + (4 * src_read_rect.size.width) as usize);
+                dest_data.push_all(row);
+                src += stride as usize;
+            }
+        }
+        // bgra -> rgba
+        byte_swap(dest_data.as_mut_slice());
+        chan.send(dest_data).unwrap();
+    }
+
+    fn put_image_data(&mut self, mut imagedata: Vec<u8>, image_data_rect: Rect<i32>,
+                      dirty_rect: Option<Rect<i32>>, canvas_size: Size2D<i32>) {
+
+        if image_data_rect.size.width <= 0 || image_data_rect.size.height <= 0 {
+            return
+        }
+
+        assert!(image_data_rect.size.width * image_data_rect.size.height * 4 == imagedata.len() as i32);
+        // rgba -> bgra
+        byte_swap(imagedata.as_mut_slice());
+
+        let new_image_data_rect = Rect(Point2D(0i32, 0i32),
+            Size2D(image_data_rect.size.width, image_data_rect.size.height));
+
+        let new_dirty_rect = match dirty_rect {
+            Some(mut dirty_rect) => {
+                if dirty_rect.size.width < 0 {
+                    dirty_rect.origin.x = dirty_rect.origin.x + dirty_rect.size.width;
+                    dirty_rect.size.width = -dirty_rect.size.width;
+                }
+                if dirty_rect.size.height < 0 {
+                    dirty_rect.origin.y = dirty_rect.origin.y + dirty_rect.size.height;
+                    dirty_rect.size.height = -dirty_rect.size.height;
+                }
+                new_image_data_rect.intersection(&dirty_rect)
+            },
+            None => Some(new_image_data_rect)
+        };
+
+        if let Some(_) = new_dirty_rect {
+            let moved_dirty_rect = Rect(new_dirty_rect.unwrap().origin.add(image_data_rect.origin),
+                                        new_dirty_rect.unwrap().size).intersection(&Rect(Point2D(0i32, 0i32),
```

Comment:
If we use `if let Some(new_dirty_rect) = new_dirty_rect {` above, we can get rid of these unwrap calls.